### PR TITLE
Add MinMaxScaler and StandardScaler methods

### LINKED ISSUE #12: MachineLearning_Toolkit: Add data preprocessing methods
#### Description:

ref here - https://scikit-learn.org/stable/modules/classes.html#module-sklearn.preprocessing

I think the MinMaxScaler, StandardScaler and Normalizer will be useful, but can add others if someone names those.

### PR DESCRIPTION
**Pull request by BingWangUS**
### Issues addressed by this PR

Closes #12 

### Test files

https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/01_Test%20Scripts/MachineLearning_Toolkit/MachineLearning%2312-PreProcessing?csf=1&web=1&e=DVzaYy

### Changelog

- Create MinMaxScaler and StandardScaler object
- Create MinMaxScaler and StandardScaler python methods
- Create MinMaxScaler and StandardScaler BHoM methods

### CODE REVIEW

File: MachineLearning_Engine/Compute/MinMaxScaler.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+        [Description("Rescale the data to a given range between 0 and 1.")]
+        [Input("x", "Data to be rescaled.")]
+        [MultiOutput(0, "scaler", "The estimated scaler for the min-max scale transformation.")]
+        [MultiOutput(1, "rescaledX", "The rescaled data with a range between o and 1.")]
+        public static Output<MinMaxScaler, Tensor> MinMaxScaler(Tensor x)
+        {
+            PyObject scaler = BH.Engine.MachineLearning.Compute.Invoke("MinMaxScaler.fit", x);
+            Tensor rescaledX = new Tensor(BH.Engine.MachineLearning.Compute.Invoke("MinMaxScaler.scale", scaler, x));
+            return new Output<MinMaxScaler, Tensor> { Item1 = new MinMaxScaler(scaler), Item2 = rescaledX };
```
Comment (epignatelli) [inline MachineLearning_Engine/Compute/MinMaxScaler.cs]: I think it would be good if we can stick to the current format of returning just the model, rather than the fitted data as well.

File: MachineLearning_Engine/Compute/MinMaxScaler.cs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        /*************************************/
+
+        [Description("Rescale the data to a given range between 0 and 1.")]
+        [Input("x", "Data to be rescaled.")]
+        [MultiOutput(0, "scaler", "The estimated scaler for the min-max scale transformation.")]
+        [MultiOutput(1, "rescaledX", "The rescaled data with a range between o and 1.")]
+        public static Output<MinMaxScaler, Tensor> MinMaxScaler(Tensor x)
+        {
+            PyObject scaler = BH.Engine.MachineLearning.Compute.Invoke("MinMaxScaler.fit", x);
+            Tensor rescaledX = new Tensor(BH.Engine.MachineLearning.Compute.Invoke("MinMaxScaler.scale", scaler, x));
```
Comment (epignatelli) [inline MachineLearning_Engine/Compute/MinMaxScaler.cs]: Can we stick to the `fit` name of the method? I.e. `MinMaxScaler.fit`?

File: MachineLearning_Engine/Compute/MinMaxScaler.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+        /*************************************/
+
+        [Description("Rescale the data with a given min-max scaler.")]
+        [Input("scaler", "The given scaler for the min-max scale transformation.")]
+        [Input("x", "Data to be rescaled.")]
+        [Output("rescaledX", "The rescaled data with a range between o and 1.")]
+        public static Tensor MinMaxScale(MinMaxScaler scaler, Tensor x)
+        {
+            return new Tensor(BH.Engine.MachineLearning.Compute.Invoke("MinMaxScaler.scale", scaler, x));
```
Comment (epignatelli) [inline MachineLearning_Engine/Compute/MinMaxScaler.cs]: Same here for `infer`. Can we go for `MinMaxScaler.infer`?

File: MachineLearning_Engine/Compute/MinMaxScaler.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        }
+
+        /*************************************/
+        [Description("Inverse transformation of data with the given scaler.")]
+        [Input("scaler", "The min-max scaler used for inverse transformation.")]
+        [Input("x", "Data to be inverse transformed.")]
+        [Output("inversedX", "The inverse transformed data using the min-max scaler.")]
+        public static Tensor InverseMinMaxScale(MinMaxScaler scaler, Tensor x)
+        {
+            return new Tensor(BH.Engine.MachineLearning.Compute.Invoke("MinMaxScaler.inverse", scaler, x));
```
Comment (epignatelli) [inline MachineLearning_Engine/Compute/MinMaxScaler.cs]: Good one. I would then go for `infer_inverse`?

File: MachineLearning_Engine/Compute/StandardScaler.cs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        /**** Public Methods              ****/
+        /*************************************/
+
+        [Description("Rescale the data with standard scaler.")]
+        [Input("x", "Data to be rescaled.")]
+        [MultiOutput(0, "scaler", "The estimated scaler for the standard scale transformation.")]
+        [MultiOutput(1, "rescaledX", "The rescaled data after standard scale transformation.")]
+        public static Output<StandardScaler, Tensor> StandardScaler(Tensor x)
+        {
+            PyObject scaler = BH.Engine.MachineLearning.Compute.Invoke("StandardScaler.fit", x);
```
Comment (epignatelli) [inline MachineLearning_Engine/Compute/StandardScaler.cs:46]: Same comment for `MinMaxScaler` applies to `StandardScaler` here.

**Review by epignatelli — CHANGES_REQUESTED** [review]
Thanks @BingWangUS ! This is great!!
I left some comments in the code, mainly format!

Can you make sure that the `ChangeLog` lists the modifications using a bulleted list?
- x
- y

File: MachineLearning_Engine/Compute/MinMaxScaler.cs (hunk 1 above)
Comment (BingWangUS) [reply MachineLearning_Engine/Compute/MinMaxScaler.cs]: Will do. I thought of this once but also think giving tensors at the same time will save one extra step for most cases. But it make sense to keep the same format.

File: MachineLearning_Engine/Compute/MinMaxScaler.cs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+            return new MinMaxScaler(scaler);
+        }
+
+        /*************************************/
+
+        [Description("Rescale the data with a given min-max scaler.")]
+        [Input("scaler", "The given scaler for the min-max scale transformation.")]
+        [Input("x", "Data to be rescaled.")]
+        [Output("rescaledX", "The rescaled data with a range between o and 1.")]
+        public static Tensor MinMaxScale(MinMaxScaler scaler, Tensor x)
```
Comment (epignatelli) [inline MachineLearning_Engine/Compute/MinMaxScaler.cs]: `public static Tensor Infer(MinMaxScaler scaler, Tensor x)` ?

File: MachineLearning_Engine/Compute/MinMaxScaler.cs [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        {
+            return new Tensor(BH.Engine.MachineLearning.Compute.Invoke("MinMaxScaler.infer", scaler, x));
+        }
+
+        /*************************************/
+        [Description("Inverse transformation of data with the given scaler.")]
+        [Input("scaler", "The min-max scaler used for inverse transformation.")]
+        [Input("x", "Data to be inverse-transformed.")]
+        [Output("inversedX", "The inverse transformed data using the min-max scaler.")]
+        public static Tensor InverseMinMaxScale(MinMaxScaler scaler, Tensor x)
```
Comment (epignatelli) [inline MachineLearning_Engine/Compute/MinMaxScaler.cs]: `public static Tensor InferInverse(MinMaxScaler scaler, Tensor x)` ?

File: MachineLearning_Engine/Compute/StandardScaler.cs [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+            return new StandardScaler(scaler);
+        }
+
+        /*************************************/
+
+        [Description("Rescale the data with a given standard scaler.")]
+        [Input("scaler", "The given scaler for the standard scale transformation.")]
+        [Input("x", "Data to be rescaled.")]
+        [Output("rescaledX", "The rescaled data after the given standard scale transformation.")]
+        public static Tensor StandardScale(StandardScaler scaler, Tensor x)
```
Comment (epignatelli) [inline MachineLearning_Engine/Compute/StandardScaler.cs]: `public static Tensor Infer(StandardScaler scaler, Tensor x)`?

File: MachineLearning_Engine/Compute/StandardScaler.cs [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        {
+            return new Tensor(BH.Engine.MachineLearning.Compute.Invoke("StandardScaler.infer", scaler, x));
+        }
+
+        /*************************************/
+        [Description("Inverse transformation of data with the given scaler.")]
+        [Input("scaler", "The standard scaler used for inverse transformation.")]
+        [Input("x", "Data to be inverse-transformed.")]
+        [Output("inversedX", "The inverse transformed data using the standard scaler.")]
+        public static Tensor InverseStandardScale(StandardScaler scaler, Tensor x)
```
Comment (epignatelli) [inline MachineLearning_Engine/Compute/StandardScaler.cs]: `public static Tensor InferInverse(StandardScaler scaler, Tensor x)`?

File: MachineLearning_Engine/Convert/ToPython.cs [hunk 10]
```diff
@@ -67,6 +67,18 @@ public static PyObject ToPython(this LinearRegression model)
             return model.SkLearnModel;
         }
 
+        /***************************************************/
+
+        public static PyObject ToPython(this MinMaxScaler scaler)
+        {
+            return scaler.SkLearnScaler;
+        }
+
```
Comment (epignatelli) [inline MachineLearning_Engine/Convert/ToPython.cs:76]: Can you add the `/***************************************************/` separator here?

**Review by epignatelli — APPROVED** [review]
Thanks @BingWangUS !!
